File: all-pairs-shortest-paths/python/all_pairs_shortest_paths_using_matrix.py
import copy
import sys


def extended_shortest_paths(L, W):
    new_L = copy.deepcopy(L)
    n = len(W)

    for i in range(n):
        for j in range(n):
            l = sys.maxsize
            for k in range(n):
                l = min(l, L[i][k] + W[k][j])

            new_L[i][j] = l

    return new_L


def all_pairs_shortest_paths(W):
    L = copy.deepcopy(W)
    n = len(W)

    for i in range(n):
        L = extended_shortest_paths(L, W)

        print(f"--- i = {i} ---")
        print_matrix(L)

    return L


def print_matrix(A):
    if not A:
        return

    for row in A:
        for i in range(len(row)):
            print(f"{row[i]:3d}" if row[i] < sys.maxsize else "INF", end=", ")
        print("")


if __name__ == "__main__":
    W = [
        [0, 3, 8, sys.maxsize, -4],
        [sys.maxsize, 0, sys.maxsize, 1, 7],
        [sys.maxsize, 4, 0, sys.maxsize, sys.maxsize],
        [2, sys.maxsize, -5, 0, sys.maxsize],
        [sys.maxsize, sys.maxsize, sys.maxsize, 6, 0],
    ]

    print("--- graph ---")
    print_matrix(W)

    L = all_pairs_shortest_paths(W)

    print("--- result ---")
    print_matrix(L)

File: all-pairs-shortest-paths/golang/all_pairs_shortest_paths_using_matrix.go
package main

import (
	"fmt"
	"math"
)

func allPairsShortestPaths(W [][]int) [][]int {
	n := len(W)
	L := make([][]int, n)
	copy(L, W)

	for i := range n {
		fmt.Printf("--- i = %d ---\n", i)

		L = extendShortestPaths(L, W)

		printMatrix(L)
	}

	return L
}

func extendShortestPaths(L, W [][]int) [][]int {
	n := len(W)
	newL := make([][]int, n)
	for i := range n {
		newL[i] = make([]int, n)
	}

	for i := range n {
		for j := range n {
			l := math.MaxInt8
			for k := range n {
				l = min(l, L[i][k]+W[k][j])
			}

			newL[i][j] = l
		}
	}

	return newL
}

func printMatrix(W [][]int) {
	for i := range W {
		for j := range W[i] {
			if W[i][j] != math.MaxInt8 {
				fmt.Printf("%3d", W[i][j])
			} else {
				fmt.Print("  -")
			}
		}

		fmt.Println("")
	}
}

func main() {
	W := [][]int{
		{0, 3, 8, math.MaxInt8, -4},
		{math.MaxInt8, 0, math.MaxInt8, 1, 7},
		{math.MaxInt8, 4, 0, math.MaxInt8, math.MaxInt8},
		{2, math.MaxInt8, -5, 0, math.MaxInt8},
		{math.MaxInt8, math.MaxInt8, math.MaxInt8, 6, 0},
	}

	fmt.Println("--- graph ---")
	printMatrix(W)

	L := allPairsShortestPaths(W)

	fmt.Println("--- result ---")
	printMatrix(L)
}
